src/Apropos/Gen.hs

```diff
@@ -181,75 +221,115 @@ type GenContext = H.Gen
+
+gStepsA :: Interleaved a -> Generator (Interleaved a)
+gStepsA m = do
+  mg <- gStep m
+  case mg of
+    Right a -> pure $ Pure a
+    Left Nothing -> pure m
+    Left (Just so) -> gStepsA so
+
+pStep :: Interleaved a -> PropertyT IO (Either (Maybe (Interleaved a)) a)
+pStep (Pure a) = pure $ Right a
+pStep (Free (P g f)) = Left . Just . f <$> g
+pStep _ = pure $ Left Nothing
+
+pSteps :: Interleaved a -> PropertyT IO (Either (Interleaved a) a)
+pSteps m = do
+  mg <- pStep m
+  case mg of
+    Right a -> pure $ Right a
+    Left Nothing -> pure $ Left m
+    Left (Just so) -> pSteps so
+
+liftG :: Generator a -> Interleaved a
+liftG g = liftF (G g id)
+
+liftP :: PropertyT IO a -> Interleaved a
+liftP p = liftF (P p id)
+
+gen2Interleaved :: Gen a -> Interleaved a
+gen2Interleaved (Free (LiftPropertyT p next)) = liftP p >>= (gen2Interleaved . next)
+gen2Interleaved (Free (Label s next)) = liftG (lift (tell (Set.singleton s))) >> gen2Interleaved next
+gen2Interleaved (Free (FailWithFootnote s _)) = liftG $ throwE $ GenException s
+gen2Interleaved (Free (GenBool next)) = liftG (lift HGen.bool) >>= (gen2Interleaved . next)
+gen2Interleaved (Free (GenInt r next)) = do
+  i <- liftG $ lift $ HGen.int (hRange r)
+  gen2Interleaved $ next i
+gen2Interleaved (Free (GenList r g next)) = do
   let gs = int r >>= \l -> replicateM l g
   gs >>>= next
-gen (Free (GenShuffle ls next)) = do
-  s <- lift $ HGen.shuffle ls
-  gen $ next s
-gen (Free (GenElement ls next)) = do
+gen2Interleaved (Free (GenShuffle ls next)) = do
+  s <- liftG $ lift $ HGen.shuffle ls
+  gen2Interleaved $ next s
+gen2Interleaved (Free (GenElement ls next)) = do
   if null ls
-    then throwE $ GenException "GenElement empty list"
+    then liftG $ throwE $ GenException "GenElement empty list"
     else do
-      s <- lift $ HGen.element ls
-      gen $ next s
-gen (Free (GenChoice gs next)) = do
+      s <- liftG $ lift $ HGen.element ls
+      gen2Interleaved $ next s
+gen2Interleaved (Free (GenChoice gs next)) = do
   let l = length gs
   if l == 0
-    then throwE $ GenException "GenChoice: list length zero"
+    then liftG $ throwE $ GenException "GenChoice: list length zero"
     else do
-      i <- lift (HGen.int (HRange.linear 0 (l - 1)))
+      i <- liftG $ lift (HGen.int (HRange.linear 0 (l - 1)))
       (gs !! i) >>>= next
-gen (Free (GenFilter c g next)) = do
-  res <- filter' c $ gen g
-  gen $ next res
-gen (Free (Scale s g next)) =
-  HGen.scale (\(HRange.Size x) -> HRange.Size (s x)) (gen g) >>= gen . next
-gen (Free (Prune g next)) = HGen.prune (gen g) >>= gen . next
-gen (Free (ThrowRetry _)) = throwE Retry
-gen (Free (OnRetry a b next)) = do
-  res <- lift $ runExceptT (gen a)
+gen2Interleaved (Free (GenFilter c g next)) = do
+  let f = do
+        res <- g
+        if c res
+          then pure res
+          else retry
+  res <- liftP $ forAllWithRetries 100 f
+  case res of
+    Left err -> liftG (throwE err)
+    Right a -> gen2Interleaved $ next a
```

Comment:
Filter is now an embedded `PropertyT IO` that uses the `forAllWithRetries` mechanism.